llvm/lib/Transforms/Scalar/InferAddressSpaces.cpp

```diff
@@ -669,17 +666,51 @@ Value *InferAddressSpacesImpl::cloneInstructionWithNewAddressSpace(
     // Technically the intrinsic ID is a pointer typed argument, so specially
     // handle calls early.
     assert(II->getIntrinsicID() == Intrinsic::ptrmask);
-    Value *NewPtr = operandWithNewAddressSpaceOrCreatePoison(
-        II->getArgOperandUse(0), NewAddrSpace, ValueWithNewAddrSpace,
-        PredicatedAS, PoisonUsesToFix);
-    Value *Rewrite =
-        TTI->rewriteIntrinsicWithAddressSpace(II, II->getArgOperand(0), NewPtr);
-    if (Rewrite) {
-      assert(Rewrite != II && "cannot modify this pointer operation in place");
-      return Rewrite;
+    const Use &PtrOpUse = II->getArgOperandUse(0);
+    unsigned OldAddrSpace = PtrOpUse.get()->getType()->getPointerAddressSpace();
+    Value *MaskOp = II->getArgOperand(1);
+    Type *MaskTy = MaskOp->getType();
+
+    bool DoTruncate = false;
+    bool DoNotConvert = false;
+
+    if (!TTI->isNoopAddrSpaceCast(OldAddrSpace, NewAddrSpace)) {
+      // All valid 64-bit to 32-bit casts work by chopping off the high
+      // bits. Any masking only clearing the low bits will also apply in the new
+      // address space.
+      if (DL->getPointerSizeInBits(OldAddrSpace) != 64 ||
+          DL->getPointerSizeInBits(NewAddrSpace) != 32) {
```

Comment:
While this is probably true for all architectures right now it seems like something that should be a TII hook.

Reply:
Is that actually true for AMDGPU `addrspacecast ptr to addrspace(3)` and `addrspace(5)`?

Reply:
that's just poison

Reply:
Wait, isn't it true that a flat pointer can point to LDS?

And so, given an LDS pointer, you can addrspacecast it back into the flat equivalent?

Reply:
https://github.com/llvm/llvm-project/blob/23f0fbf8fff563c77f770f83096b522c3c99a82d/llvm/lib/Target/AMDGPU/AMDGPU.h#L561

seems to suggest that, too, I think. This function is called by isValidAddrSpaceCast in AMDGPUTargetTransformInfo:
https://github.com/llvm/llvm-project/blob/23f0fbf8fff563c77f770f83096b522c3c99a82d/llvm/lib/Target/AMDGPU/AMDGPUTargetTransformInfo.h#L180-L187

Reply:
> While this is probably true for all architectures right now it seems like something that should be a TTI hook.

done, btw

Reply:
Flat to LDS works. LDS to private does not work, and is poison